# Ejercicio de Segmentación de clientes

Empezamos trayéndonos el fichero excel con las ventas online. 
Como os puede dar 'warnings' podéis ayudaros de la librería warnings y filtrarlos para ignorarlos.
Usaremos el dataset 'Online_Retail.xlsx'

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

Como lleva unos minutos cargar los datos, guardamos una copia como copia de seguridad.

Tenemos que revisar los datos de forma que conozcamos su distribución

38

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

Country  CustomerID
36        United Kingdom        3950
14               Germany          95
13                France          87
31                 Spain          31
3                Belgium          25
33           Switzerland          21
27              Portugal          19
19                 Italy          15
12               Finland          12
1                Austria          11
25                Norway          10
24           Netherlands           9
0              Australia           9
6        Channel Islands           9
9                Denmark           9
7                 Cyprus           8
32                Sweden           8
20                 Japan           8
26                Poland           6
34                   USA           4
5                 Canada           4
37           Unspecified           4
18                Israel           4
15                Greece           4
10                  EIRE           3
23                 Malta           2
35  United Arab Emirates           2
2                Bahrain           2
22             Lithuania           1
8         Czech Republic           1
21               Lebanon           1
28                   RSA           1
29          Saudi Arabia           1
30             Singapore           1
17               Iceland           1
4                 Brazil           1
11    European Community           1
16             Hong Kong           0

Más del 90% de los clientes en los datos son de Reino Unido. Hay algunas investigaciones que indican que los clústeres de clientes varían según la geografía, así que aquí restringiré los datos a Reino Unido solamente.

Comprobamos si faltan valores en cada columna.

Faltan 133600 valores en la columna CustomerID, ya que nuestro análisis se basa en clientes, eliminaremos estos valores que faltan.

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133600
Country             0
dtype: int64

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

Comprueba los valores mínimo y máximo en la columna UnitPrice

0.0

-80995

Eliminamos los valores negativos de la columna Quantity

1

Después de limpiar, ahora tratamos con 354345 filas y 8 columnas. Verlo y comprobar con info()

(354345, 8)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 354345 entries, 0 to 541893
Data columns (total 8 columns):
InvoiceNo      354345 non-null object
StockCode      354345 non-null object
Description    354345 non-null object
Quantity       354345 non-null int64
InvoiceDate    354345 non-null datetime64[ns]
UnitPrice      354345 non-null float64
CustomerID     354345 non-null float64
Country        354345 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 24.3+ MB


Compruebe el valor único para cada columna. Crearos una función que recorra todas las columnas y ponga identificador : contador.
Después echar un vistazo a la cabecera.

InvoiceNo :  16649
StockCode :  3645
Description :  3844
Quantity :  294
InvoiceDate :  15615
UnitPrice :  403
CustomerID :  3921
Country :  1


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

Añadimos una columna para el precio total

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom       15.30  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom       22.00  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34

Descubrimos la primera y última fecha de pedido en los datos

Timestamp('2010-12-01 08:26:00')

Timestamp('2011-12-09 12:49:00')

Para ver la recencia (días transcurridos desde la última compra) se calcula para un punto en el tiempo. La última fecha de la factura es 2011-12-09, esta es la fecha que usaremos para calcularlo.

In [ ]:
import datetime as dt


Creamos una tabla **RFM** (Recency, Frecuency, Money):
- **Recencia.** Días transcurridos desde la última compra.
- **Frecuencia.** Número de compras por período de tiempo, como promedio. Por ejemplo, número de compras mensuales.
- **Money.** Valor de las compras totales realizadas por el cliente en el tiempo de análisis.

Calculamos métricas RFM para cada cliente

recency  frequency  monetary_value
CustomerID                                    
12346.0         325          1        77183.60
12747.0           2        103         4196.01
12748.0           0       4596        33719.73
12749.0           3        199         4090.88
12820.0           3         59          942.34

Interpretación:

- CustomerID 12346 tiene frecuencia:1, valor monetario: $77183.60 y recencia:324 días.

- CustomerID 12747 tiene frecuencia: 103, valor monetario: $4196.01 y recencia: 1 día

Vamos a comprobar los detalles del primer cliente.

InvoiceNo StockCode                     Description  Quantity  \
61619    541431     23166  MEDIUM CERAMIC TOP STORAGE JAR     74215   

              InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
61619 2011-01-18 10:01:00       1.04     12346.0  United Kingdom     77183.6

El primer cliente ha comprado sólo una vez: compró un artículo en una gran cantidad (74215). El precio unitario es muy bajo, parece que viene de una liquidación.

True

La forma de dividir las métricas en segmentos es mediante el uso de cuartil. 

1. Esto nos da un punto de partida para un análisis detallado
2. 4 segmentos son fáciles de entender y explicar

recency  frequency  monetary_value
0.25     17.0       17.0          300.04
0.50     50.0       41.0          651.82
0.75    142.0       99.0         1575.89

{'recency': {0.25: 17.0, 0.5: 50.0, 0.75: 142.0},
 'frequency': {0.25: 17.0, 0.5: 41.0, 0.75: 99.0},
 'monetary_value': {0.25: 300.03999999999996,
  0.5: 651.8199999999999,
  0.75: 1575.89}}

Creamos una tabla RFM segmentada

La más baja recencia, mayor frecuencia y dinero son nuestros mejores clientes 

Añadimos números de segmento a nuestra tabla RFM

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12747.0           2        103         4196.01           1           1   
12748.0           0       4596        33719.73           1           1   
12749.0           3        199         4090.88           1           1   
12820.0           3         59          942.34           1           2   

            m_quartile  
CustomerID              
12346.0              1  
12747.0              1  
12748.0              1  
12749.0              1  
12820.0              2

Los segmentos RFM dividen su base de clientes en un cubo 3D imaginario. Es difícil de visualizar. Sin embargo, podemos resolverlo.

Agrega una nueva columna para combinar la puntuación RFM, 111 es la puntuación más alta como determinó anteriormente.

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12747.0           2        103         4196.01           1           1   
12748.0           0       4596        33719.73           1           1   
12749.0           3        199         4090.88           1           1   
12820.0           3         59          942.34           1           2   

            m_quartile RFMScore  
CustomerID                       
12346.0              1      441  
12747.0              1      111  
12748.0              1      111  
12749.0              1      111  
12820.0              2      122

Aparentemente, el primer cliente no es nuestro mejor cliente en absoluto.

¡Aquí está el top 10 de nuestros mejores clientes!

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
18102.0           0        431       259657.30           1           1   
17450.0           8        337       194550.79           1           1   
17511.0           2        963        91062.38           1           1   
16684.0           4        277        66653.56           1           1   
14096.0           4       5111        65164.79           1           1   
13694.0           3        568        65039.62           1           1   
15311.0           0       2379        60767.90           1           1   
13089.0           2       1818        58825.83           1           1   
15769.0           7        130        56252.72           1           1   
15061.0           3        403        54534.14           1           1   

            m_quartile RFMScore  
CustomerID                       
18102.0              1      111  
17450.0              1      111  
17511.0              1      111  
16684.0              1      111  
14096.0              1      111  
13694.0              1      111  
15311.0              1      111  
13089.0              1      111  
15769.0              1      111  
15061.0              1      111

Para profundizar ver en:

[Buy ’Til You Die - A Walkthrough](https://cran.r-project.org/web/packages/BTYD/vignettes/BTYD-walkthrough.pdf)

[Customer Segmentation for a wine seller](http://blog.yhat.com/posts/customer-segmentation-using-python.html)

Referencia: [Blast Analytics and Marketing](http://www.blastam.com/blog/rfm-analysis-boosts-sales)